client/lobby/SelectionTab.cpp

```diff
@@ -328,29 +345,82 @@ void SelectionTab::showPopupWindow(const Point & cursorPosition)
 	if(py >= curItems.size())
 		return;
 
-	std::string text = boost::str(boost::format("{%1%}\r\n\r\n%2%:\r\n%3%") % curItems[py]->getName() % CGI->generaltexth->translate("vcmi.lobby.filename") % curItems[py]->fileURI);
-	if(curItems[py]->date != "")
-	    text += boost::str(boost::format("\r\n\r\n%1%:\r\n%2%") % CGI->generaltexth->translate("vcmi.lobby.creationDate") % curItems[py]->date);
+	if(!curItems[py]->isFolder)
+	{
+		std::string text = boost::str(boost::format("{%1%}\r\n\r\n%2%:\r\n%3%") % curItems[py]->getName() % CGI->generaltexth->translate("vcmi.lobby.filename") % curItems[py]->fileURI);
+		if(curItems[py]->date != "")
+			text += boost::str(boost::format("\r\n\r\n%1%:\r\n%2%") % CGI->generaltexth->translate("vcmi.lobby.creationDate") % curItems[py]->date);
 
-	CRClickPopup::createAndPush(text);
+		CRClickPopup::createAndPush(text);
+	}
+}
+
+std::tuple<std::string, std::string, bool, bool> SelectionTab::checkSubfolder(std::string path)
+{
+	std::string folderName = "";
+	bool parentExists = (curFolder != "");
+	bool fileInFolder = false;
+
+	std::vector<std::string> filetree;
+	// delete first element (e.g. 'MAPS')
+	boost::split(filetree, path, boost::is_any_of("/"));
+	filetree.erase(filetree.begin());
+	std::string pathWithoutPrefix = boost::algorithm::join(filetree, "/");
+
+	std::string baseFolder = boost::filesystem::path(pathWithoutPrefix).parent_path().string();
+
+	if(boost::algorithm::starts_with(baseFolder, curFolder))
+	{
+		std::string folder = baseFolder.substr(curFolder.size());
+
+		if(folder != "")
+		{
+			boost::split(filetree, folder, boost::is_any_of("/"));
+			folderName = filetree[0];
+		}
+	}
+
+	if(boost::algorithm::starts_with(pathWithoutPrefix, curFolder))
+		if(boost::count(pathWithoutPrefix.substr(curFolder.size()), '/') == 0)
+			fileInFolder = true;
+
+    return {folderName, baseFolder, parentExists, fileInFolder};
 }
 
 // A new size filter (Small, Medium, ...) has been selected. Populate
 // selMaps with the relevant data.
 void SelectionTab::filter(int size, bool selectFirst)
 {
+	if(size == -1)
+		size = curFilterSize;
+	curFilterSize = size;
+
 	curItems.clear();
 
-	if(tabType == ESelectionScreen::campaignList)
-	{
-		for(auto elem : allItems)
-			curItems.push_back(elem);
-	}
-	else
+	for(auto elem : allItems)
 	{
-		for(auto elem : allItems)
+		if((elem->mapHeader && (!size || elem->mapHeader->width == size)) || tabType == ESelectionScreen::campaignList)
 		{
-			if(elem->mapHeader && (!size || elem->mapHeader->width == size))
+			auto [folderName, baseFolder, parentExists, fileInFolder] = checkSubfolder(elem->fileURI);
+
+			if(parentExists)
+			{
+				auto folder = std::make_shared<ElementInfo>();
+				folder->isFolder = true;
+				folder->folderName = "..";
+				if (boost::range::find_if(curItems, [](std::shared_ptr<ElementInfo> e) { return e->folderName == ".."; }) == curItems.end()) {
+					curItems.push_back(folder);
+				}				
```

Comment:
Perhaps split into two statements for clarity?
```suggestion
				auto itemIt = boost::range::find_if(curItems, [](std::shared_ptr<ElementInfo> e) { return e->folderName == ".."; })
				if (itemIt == curItems.end()) {
					curItems.push_back(folder);
				}				
```